grails-app/services/org/pih/warehouse/inventory/InventorySnapshotService.groovy

```diff
@@ -464,155 +464,207 @@ class InventorySnapshotService {
-                    quantityInbound +
-                    quantityOutbound
-
-            def row = [
-                    "Code"          : product.productCode,
-                    "Name"          : product.displayNameWithLocaleCode,
-                    "Product Family": product?.productFamily?.name ?: '',
-                    "Category"      : product.category.name,
-                    "Formulary"     : product.productCatalogsToString(),
-                    "Tag"           : product.tagsToString(),
-                    "Unit Cost"     : product.pricePerUnit ?: ''
-            ]
-            row.put("Opening", balanceOpening)
-            def includeRow = balanceOpening || balanceClosing || quantityAdjustments
-            transactionTypes.each { TransactionType transactionType ->
-                String translatedTransactionTypeName = transactionType?.name
-                if (transactionType?.name?.contains(Constants.LOCALIZED_STRING_SEPARATOR)) {
-                    translatedTransactionTypeName = LocalizationUtil.getLocalizedString(transactionType?.name)
+                if (orderBy) {
+                    order(orderBy, 'desc')
                 }
-                def quantity =
-                        transactionData.find {
-                            it.productCode == product.productCode && transactionType.compareName(it?.transactionTypeName)
-                        }?.quantity?:0
-                row[translatedTransactionTypeName] = quantity
-                includeRow = quantity != 0 ? true : includeRow
             }
+        } as List<TransactionEntry>
+    }
 
-            row.put("Adjustments", quantityAdjustments)
-            row.put("Closing", balanceClosing)
-
-            if (includeRow) {
-                data << row
-            }
+    Map<Product, Map<String, Integer>> getDetailedTransactionReportData(Map<Product, List<TransactionEntry>> transactionEntries) {
+        return transactionEntries.collectEntries { product, entriesForProduct ->
+            Map<String, Integer> totalsByType = entriesForProduct
+                    .groupBy { entry ->
+                        entry.transaction.transactionType.name
+                    }
+                    .collectEntries { transactionTypeName, entriesByType ->
+                        Integer total = entriesByType.sum { entry ->
+                            Math.abs(entry.quantity as Integer)
+                        }
+                        [(transactionTypeName): total]
+                    }
+
+            [(product): totalsByType]
         }
-        return data.sort { it."Code" }
     }
 
-    def getTransactionReportSummary(Location location, List<Category> categories, List<Tag> tags, List<ProductCatalog> catalogs, Date startDate, Date endDate) {
+    List<Object> getTransactionReport(Location location, List<Category> categories, List<Tag> tagsList, List<ProductCatalog> catalogsList, Date startDate, Date endDate, Boolean includeDetails) {
```

Comment:
it's a bit hard to tell with all the changes, but this flow is all newly added by you, right? If so, I think this would be better off in a different service since it doesn't actually have anything to do with inventory snapshots anymore and we're starting to deprecate/remove all inventory snapshot stuff. You could even put it in a new TransactionReportService.